x/evm/keeper/msg_server.go

```diff
@@ -143,21 +143,20 @@ func (s msgServer) VoteConfirmGatewayDeployment(c context.Context, req *types.Vo
 
 	keeper := s.ForChain(chain.Name)
 
-	if _, ok := keeper.GetGatewayAddress(ctx); ok {
-		return &types.VoteConfirmGatewayDeploymentResponse{Log: "gateway is already confirmed"}, nil
-	}
-
-	address, ok := keeper.GetPendingGatewayAddress(ctx)
-	if !ok {
-		return nil, fmt.Errorf("no pending gateway found")
-	}
-
 	voter := s.snapshotter.GetOperator(ctx, req.Sender)
 	if voter == nil {
 		return nil, fmt.Errorf("account %v is not registered as a validator proxy", req.Sender.String())
 	}
 
 	poll := s.voter.GetPoll(ctx, req.PollKey)
+	switch {
+	case poll.Is(vote.Expired):
+		return &types.VoteConfirmGatewayDeploymentResponse{Log: fmt.Sprintf("vote for poll %s already expired", req.PollKey)}, nil
+	case poll.Is(vote.Failed), poll.Is(vote.Completed):
```

Comment:
This can become some potential DoS though. Shouldn't we check if the voter already voted instead?

Reply:
good point

Reply:
how can it DoS? poll.Vote is going to fail if you are not allowed to vote

Reply:
makes sense. I am feeling like a weathercock 😄